File: System-Design/Cryptography/Merkle-Tree/merkleTree.ts
import { createHash } from 'crypto';

export class MerkleTree {
    private leaves: string[];
    private tree: string[][] = [];
    private root: string;

    constructor(dataBlocks: string[]) {
        this.leaves = dataBlocks.map(block => this.hash(block));
        this.root = this.buildTree();
    }

    private hash(data: string): string {
        return createHash('sha256').update(data).digest('hex');
    }

    private buildTree(): string {
        if (this.leaves.length === 0) return "";

        let currentLevel = this.leaves;
        this.tree.push(currentLevel);

        while (currentLevel.length > 1) {
            const nextLevel: string[] = [];
            for (let i = 0; i < currentLevel.length; i += 2) {
                const left = currentLevel[i];
                // Duplicate last node if odd
                const right = (i + 1 < currentLevel.length) ? currentLevel[i + 1] : left;
                nextLevel.push(this.hash(left + right));
            }
            this.tree.push(nextLevel);
            currentLevel = nextLevel;
        }

        return currentLevel[0];
    }

    public getRoot(): string {
        return this.root;
    }
}

// --- CI/CD Automated Test ---
const txs = ["Block 1", "Block 2", "Block 3"];
const treeOriginal = new MerkleTree(txs);

const txsHacked = ["Block 1", "Block HACKED", "Block 3"];
const treeHacked = new MerkleTree(txsHacked);

if (treeOriginal.getRoot() !== treeHacked.getRoot()) {
    console.log("TypeScript Merkle Tree Test Passed!");
} else {
    process.exit(1);
}

File: System-Design/Cryptography/Merkle-Tree/merkle_tree.py
import hashlib

class MerkleTree:
    def __init__(self, data_blocks):
        self.leaves = [self._hash(block) for block in data_blocks]
        self.tree = []
        self.root = self._build_tree()

    def _hash(self, data: str) -> str:
        return hashlib.sha256(data.encode('utf-8')).hexdigest()

    def _build_tree(self):
        if not self.leaves:
            return ""
        
        current_level = self.leaves
        self.tree.append(current_level)

        while len(current_level) > 1:
            next_level = []
            for i in range(0, len(current_level), 2):
                left = current_level[i]
                # Duplicate the last node if there is an odd number
                right = current_level[i + 1] if i + 1 < len(current_level) else left
                next_level.append(self._hash(left + right))
            
            self.tree.append(next_level)
            current_level = next_level

        return current_level[0]

    def get_root(self):
        return self.root

# --- CI/CD Automated Test ---
if __name__ == '__main__':
    # Simulating 3 transactions
    transactions = ["Tx1: Alice pays Bob 5 BTC", "Tx2: Bob pays Charlie 2 BTC", "Tx3: Charlie pays Dave 1 BTC"]
    
    tree1 = MerkleTree(transactions)
    root1 = tree1.get_root()
    
    # Simulating a hacker altering a single transaction
    hacked_transactions = ["Tx1: Alice pays Bob 5 BTC", "Tx2: Bob pays Hacker 2 BTC", "Tx3: Charlie pays Dave 1 BTC"]
    tree2 = MerkleTree(hacked_transactions)
    root2 = tree2.get_root()
    
    assert root1 != root2, "Merkle Root failed to detect data tampering!"
    assert len(root1) == 64, "SHA-256 hash length is incorrect!"
    
    print("Python Merkle Tree Test Passed!")